core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +321,104 @@ private void balance() {
             balanceByWeights();
         }
 
+        /**
+         * Makes a decision about moving a single shard to a different node to form a more
+         * optimally balanced cluster.  This method is invoked from the cluster allocation
+         * explain API only.
+         */
+        private RebalanceDecision decideRebalance(final ShardRouting shard) {
+            if (shard.started() == false) {
+                // cannot rebalance a shard that isn't started
+                return RebalanceDecision.NOT_TAKEN;
+            }
+
+            Decision canRebalance = allocation.deciders().canRebalance(shard, allocation);
+            if (canRebalance.type() != Type.YES) {
+                // pick the first NO rebalance decision and use its explanation for the final explanation
+                String explanation = null;
+                Type explanationCause = null;
+                for (Decision subDecision : canRebalance.getDecisions()) {
+                    if ((subDecision.type() == Type.NO && (explanation == null || explanationCause == Type.THROTTLE))
+                            || (subDecision.type() == Type.THROTTLE && explanation == null)) {
+                        explanation = subDecision.getExplanation();
+                        explanationCause = subDecision.type();
+                    }
+                }
+                return new RebalanceDecision(canRebalance, Type.NO, explanation);
+            }
+
+            if (allocation.hasPendingAsyncFetch()) {
+                return new RebalanceDecision(
+                    canRebalance,
+                    Type.NO,
+                    "cannot rebalance due to in-flight shard store fetches, otherwise allocation may prematurely rebalance a shard to " +
+                        "a node that is soon to receive another shard assignment upon completion of the shard store fetch, " +
+                        "rendering the cluster imbalanced again"
+                );
+            }
+
+            sorter.reset(shard.getIndexName());
+            ModelNode[] modelNodes = sorter.modelNodes;
+            final String currentNodeId = shard.currentNodeId();
+            // find currently assigned node
+            ModelNode currentNode = null;
+            for (ModelNode node : modelNodes) {
+                if (node.getNodeId().equals(currentNodeId)) {
+                    currentNode = node;
+                    break;
+                }
+            }
+            assert currentNode != null : "currently assigned node could not be found";
+
+            // balance the shard, if a better node can be found
+            final float currentWeight = sorter.weight(currentNode);
+            final AllocationDeciders deciders = allocation.deciders();
+            final String idxName = shard.getIndexName();
+            Map<String, NodeRebalanceDecision> nodeDecisions = new HashMap<>();
```

Comment:
minor optimization, but we can do `new HashMap<>(modelNodes.length - 1);` to avoid hashmap resizing (if you want)